frameworks.qmd

```diff
@@ -1,119 +1,1785 @@
+edge deployment. We will compare programming models, supported hardware,
+optimization capabilities, and more to fully understand how frameworks
+enable scalable machine learning from the cloud to the edge.
+
+## Framework Evolution
+
+Machine learning frameworks have evolved significantly over time to meet
+the diverse needs of machine learning practitioners and advancements in
+AI techniques. A few decades ago, building and training machine learning
+models required extensive low-level coding and infrastructure. Machine
+learning frameworks have evolved considerably over the past decade to
+meet the expanding needs of practitioners and rapid advances in deep
+learning techniques. Early neural network research was constrained by
+insufficient data and compute power. Building and training machine
+learning models required extensive low-level coding and infrastructure.
+But the release of large datasets like ImageNet in 2009 and advancements
+in parallel GPU computing unlocked the potential for far deeper neural
+networks.
+
+The first ML frameworks, Theano (2007) and Caffe (2014), were developed
+by academic institutions (Montreal Institute for Learning Algorithms,
+Berkeley Vision and Learning Center). Amid a growing interest in deep
+learning due to state-of-the-art performance of AlexNet (2012) on the
+ImageNet dataset, private companies and individuals began developing ML
+frameworks, resulting in frameworks such as Keras by Google researcher
+François Chollet (2015), Chainer by Preferred Networks (2015),
+TensorFlow by Google (2015), CNTK by Microsoft (2016), and PyTorch by
+Facebook (2016).
+
+Many of these ML frameworks can be divided into categories, namely
+high-level vs. low-level frameworks and static vs. dynamic computational
+graph frameworks. High-level frameworks provide a higher level of
+abstraction than low-level frameworks. That is, high-level frameworks
+have pre-built functions and modules for common ML tasks, such as
+creating, training, and evaluating common ML models as well as
+preprocessing data, engineering features, and visualizing data, which
+low-level frameworks do not have. Thus, high-level frameworks may be
+easier to use, but are not as customizable as low-level frameworks (i.e.
+users of low-level frameworks can define custom layers, loss functions,
+optimization algorithms, etc.). Examples of high-level frameworks
+include TensorFlow/Keras and PyTorch. Examples of low-level ML
+frameworks include TensorFlow with low-level APIs, Theano, Caffe,
+Chainer, and CNTK.
+
+Frameworks like Theano and Caffe used static computational graphs which
+required rigidly defining the full model architecture upfront. Static
+graphs require upfront declaration and limit flexibility. Dynamic graphs
+construct on-the-fly for more iterative development. But around 2016,
+frameworks began adopting dynamic graphs like PyTorch and TensorFlow 2.0
+which can construct graphs on-the-fly. This provides greater flexibility
```

Comment:
"Dynamic graphs
construct on-the-fly for more iterative development. But around 2016,
frameworks began adopting dynamic graphs like PyTorch and TensorFlow 2.0
which can construct graphs on-the-fly." 

These sentences are difficult to read, I would suggest reversing them + some tweaks: "But around 2016, frameworks such as Pytorch and TF 2.0 began adopting dynamic graphs which are constructed on-the-fly for more iterative development."

Reply:
Thanks for providing this suggestion on how to rework things. That makes things so much easier to integrate.